torch/lib/c10d/FileStore.cpp

```diff
@@ -0,0 +1,284 @@
+#include "FileStore.hpp"
+
+#include <sys/file.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <sstream>
+#include <system_error>
+#include <thread>
+
+#define SYSASSERT(rv) \
+  if ((rv) < 0) throw std::system_error(errno, std::system_category());
+
+#define SYSCHECK(expr) {                                                \
+    errno = 0; auto ___output = (expr); (void)___output;                \
+    SYSASSERT(errno);                                                   \
+  }
+
+namespace c10d {
+
+namespace {
+
+template<typename F>
+typename std::result_of<F()>::type syscall(F fn) {
+  while (true) {
+    auto rv = fn();
+    if (rv == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+    }
+    return rv;
+  }
+}
+
+// For a comprehensive overview of file locking methods,
+// see: https://gavv.github.io/blog/file-locks/.
+// We stick to flock(2) here because we don't care about
+// locking byte ranges and don't want locks to be process-wide.
+
+// RAII wrapper around flock(2)
+class Lock {
+ public:
+  explicit Lock(int fd, int operation) : fd_(fd) {
+    flock(operation);
+  }
+
+  ~Lock() {
+    unlock();
+  }
+
+  Lock(const Lock& that) = delete;
+
+  Lock(Lock&& other) noexcept {
+    fd_ = other.fd_;
+    other.fd_ = -1;
+  }
+
+  void unlock() {
+    if (fd_ >= 0) {
+      flock(LOCK_UN);
+      fd_ = -1;
+    }
+  }
+
+ protected:
+  int fd_;
+
+  void flock(int operation) {
+    auto rv = syscall(std::bind(::flock, fd_, operation));
```

Comment:
nit: this can be just wrapped in `SYSASSERT`

Reply:
True, but sticking to a single pattern here is a bit clearer IMO.